### PR TITLE
Update app-model documentation

### PR DESCRIPTION
**Pull request by lucyleeow**
# Description

Updates the app-model documentation:

* Fleshes out existing information and adds napari specific information.
* Add guide on testing app-model
* Add details on app-model vs action manager implementation differences

This depends on napari 4991 as I link to a file in the repo that is created in that PR.

### CODE REVIEW

**Comment by lucyleeow** [comment]
ping @DragaDoncila @jni who may be interested

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md:6]: @lucyleeow I think it's probably still good to keep this notice here? Afaik there's no public interface for the app?

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
reusable by any Python GUI application).
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
and menus that make up the application.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
  [menus and submenus](appp-model-menus) that have been registered with the application.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
  between a [KeyBinding](app-model-keybindings) and a command ID in the
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
The `app-model`
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
page provides a good general introduction to the app model.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
The {class}`app_model.types.Action` class is designed to easily create a high level
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
declaration of `Action`s.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
# `layers` will be injected later when this action is invoked
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
is essentially a shortcut for registering objects with the following registries:
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
currently `CommandID` enums but we are considering changing to using
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md:167]:
```suggestion
`enablement` determines whether the command is visible in the UI. It does not prevent
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
you to use a callable (instead of an expression) to determine toggle state.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
Commands are usually defined with their Action definition or with its class,
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
and how they are ordered within each section. See {class}`~app_model.types.MenuItemBase` for
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
to a "close window" action, you likely want to close the *current* window.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md:322]:
```suggestion
on the type annotations in the command function definition. A `provider`, a function
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
to the viewer. It is performed by `processors`, functions that accept an instance of a
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
and processors in the `Store` will automatically be used when executing
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
in the callback function's {py:attr}`function.__globals__` attribute.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
In the napari `app`s `Store`, some basic napari objects are added to the `namespace`
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
This means that when annotating a callback function, you will not need to import
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
can cause conflicts for subsequent tests, when running a suite of tests.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
it so a new instance of `app` is returned every time {func}`napari._app_model.get_app`
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
is used inside a test. Note that a new `app` instance is also created when the `make_napari_viewer` fixture is used.
```

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [inline docs/developers/architecture/app_model.md]:
```suggestion
Qt items are not registered because we did not think it would be best practice to
```

**Review by DragaDoncila — COMMENTED** [review]
@lucyleeow thanks for getting started on this - so much useful information in here. I know it's still in draft but I've done a first pass and left mostly grammar/spelling fixes as well as a couple of other suggestions. Feel free to ignore if you have a different workflow!

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:6]: No, but it is now inside the 'architecture' section which only includes docs aimed at devs, so we can infer that it is not part of the interface. I guess it's good to spell it out.

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: Yeah I was not sure on this one. I think @jni said no backticks for napari, and I guess I thought it would be weird if napari package didn't get backticks but app-model did...? WDYT?

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:167]: Nah `when` controls visibility. A visible, disabled action appears gray-ed out on the menu.

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:322]: hmm not sure about this one. When should we use `'` and when should we use backtick for provider/processor ?

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: `get_app` is called in `init_qactions` which is called in `make_napari_viewer` so yeah you're right. It is not returned though.

Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: So it seems that `_mock_app` is called during setup of any test (meaning a `app` is created for every test?). Is this because of the `autouse` ?

**Comment by lucyleeow** [comment]
Okay I think this is ready for review now. Thanks @DragaDoncila for the thorough review already!

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [inline docs/developers/architecture/app_model.md]: This link gives me a 404, was there a reason it's that instead of the general one:
https://app-model.readthedocs.io/en/latest/

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [reply docs/developers/architecture/app_model.md]: I think I'd prefer app-model in backticks to stand out more as something different than normal words.

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [reply docs/developers/architecture/app_model.md]: I'd put this mention of app-model in backticks too, see below.

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [inline docs/developers/architecture/app_model.md]: This link is also borked, see above.

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [inline docs/developers/architecture/app_model.md]:
```suggestion
'Action' object that is the "complete" representation of a command: a pointer
```

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [inline docs/developers/architecture/app_model.md]:
```suggestion
In napari, non-Qt `Action`s are defined in
```

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [inline docs/developers/architecture/app_model.md:153]: Worth putting in an `important` admonition?

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [reply docs/developers/architecture/app_model.md:322]: I generally think of backticks for Python code, so function/method/variable/attribute names etc. (this is why I would put `app-model` and `in-and-out` in backticks, code-style)
" would then be used to indicate/emphasize a specific word being used in (or out of) a specific context vs. necessarily it's dictionary meaning. I'd avoid ' personally.

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [inline docs/developers/architecture/app_model.md]:
```suggestion
menus {class}`~app_model.types.MenuRule`'s `when` field (see [](app-model-menus)
```

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [inline docs/developers/architecture/app_model.md]: > actions are defined executed when

This is confusing to me, either one of those should be dropped (executed?) or there is an "and" missing?

**Review by psobolewskiPhD — CHANGES_REQUESTED** [review]
Sorry this fell through the cracks!
It's really good and useful, I did catch a few broken links, so requesting changes, but otherwise it's good to go IMO.

File: docs/developers/architecture/app_model.md
Comment (Czaki) [reply docs/developers/architecture/app_model.md:6]: But as I understand, there is intention to make it public:
https://github.com/napari/napari/issues/6054

Without making get_app public, the actions cannot be simply called from plugins.

Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:6]: If we do make it public we can amend this message.

I think the intention is more to allow users to register actions at runtime. Depending on the contribution type, the user may already have a public API func to access plugin actions- e.g., there is the `open_sample` func for sample contributions, and I think there may be a way for the user to get reader/writers too?

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: (yes it is due to `autouse` that `_mock_app` fixture is called during setup of every test)

File: docs/developers/architecture/app_model.md
Comment (Czaki) [reply docs/developers/architecture/app_model.md:6]: I think that option to call actions is as important, as runtime registration

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: No reason, I think I've had the app model docs open in a tab for a long time and theres been a new release since and when I copied the URL it was to an older version? No idea, I will fix!

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: Oh yeah I mucked that up. I think I mean 'executed'? (because define just means which file they are written in right??). When a function is connected to an event or registered to the `app`, the function itself is 'run' right? I mean we import the e.g., view menu actions and that import runs the functions inside the file where the view menu actions are defined?!

What is the best phrasing here?

cc @DragaDoncila

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:153]: Good idea!

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:322]: Sorry to clarify, you would use `"provider"` instead of `'provider'` ?

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:6]: You think users should call a command in `app` via the command ID string instead of using existing public API like `open_sample`? Why?
Or do you just mean you want users to have the option of calling a command via string?

File: docs/developers/architecture/app_model.md
Comment (psobolewskiPhD) [reply docs/developers/architecture/app_model.md:322]: If provider is not a Python "thing", I'd use "

Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:322]: Done!

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: I've added a note about this.

File: docs/developers/architecture/app_model.md
Comment (Czaki) [reply docs/developers/architecture/app_model.md:6]: Because we provide public API only for opening sample and adding plugin widget. And plugins could provide many more contributions. 

And there is an area for people who would like to provide custom controllers (either widget or hardware)

Comment (lucyleeow) [reply docs/developers/architecture/app_model.md:6]: Sure but these are not the current comtribution types? We could create public API for new contribution types. Have we already decided how these new contributions would look?

I'm not opposed to execute via app model but there's probably a few decisions to be made here and this way may not be the best?

**Review by psobolewskiPhD — APPROVED** [review]
Looks good to me, I'll let @DragaDoncila & @Czaki make the final call on the two open questions.

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [reply docs/developers/architecture/app_model.md:6]: Yeah I don't think this is going to get hashed out in this PR lol.

Let's update the message to: This is not currently a part of the public napari API, but we are considering ways to expose action execution directly. 

Once we actually decide what the interface for that is, we will amend.

File: docs/developers/architecture/app_model.md
Comment (DragaDoncila) [reply docs/developers/architecture/app_model.md]: I would not use `executed` in this context. I would say "defined" when referring to the file they are literally written in, and `registered` when they are added to the app-model/action_manager registry.

**Review by DragaDoncila — COMMENTED** [review]
Left a few comments to unresolved conversations. @lucyleeow ping me once these are done and I'll approve!

File: docs/developers/architecture/app_model.md
Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: Yeah good point. I've remembered what I was thinking about when I wrote this -> plugins menu actions used to be registered later when we create the plugins menu (when we know we have Qt). Now we register plugin actions much earlier, when we don't know if we have Qt. (Thus we've had to do 'safe register')

Comment (lucyleeow) [reply docs/developers/architecture/app_model.md]: Done @DragaDoncila, let me know what you think, thanks!

**Review by DragaDoncila — APPROVED** [review]
Looks good to me! Let's get this in and we can always improve in future PRs. Thanks @lucyleeow :pray:

**Comment by lucyleeow** [comment]
@psobolewskiPhD could this go in? :pray: I've got a new PR that depends on it :grimacing:

**Pull request merged**
